### PR TITLE
Extension methods can be exported by name

### PR DESCRIPTION
**Pull request by Akirathan**
First part of #10258 - extension methods can be exported by name

### Pull Request Description

Ultimately, we want to forbid the `from ... export all` syntax. This PR starts by providing a way to explicitly export extension and conversion methods by name.

Stdlib code will be modified in upcoming PR.

### Important Notes

A single name can refer to multiple extension or conversion methods. Exports are not qualified. For example, 
```
type My_Type
type Other_Type
My_Type.ext_method x = x
Other_Type.ext_method x = x
```
```
from project.Mod export ext_method
```
will export both `My_Type.ext_method` and `Other_Type.ext_method`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: lib/java/test-utils/src/main/java/org/enso/test/utils/ModuleUtils.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    return getExportedSymbols(mod);
+  }
+
+  public static List<DefinedEntity> getDefinedEntities(Context ctx, String modName) {
+    var ensoCtx = ContextUtils.leakContext(ctx);
+    var mod = ensoCtx.getPackageRepository().getLoadedModule(modName).get();
+    return CollectionConverters.asJava(mod.getBindingsMap().definedEntities());
+  }
+
+  private static Map<String, List<ResolvedName>> getExportedSymbols(Module module) {
```
Comment (JaroslavTulach) [inline lib/java/test-utils/src/main/java/org/enso/test/utils/ModuleUtils.java:35]: This method looks almost like a useful production code. `module.getExportedSymbols` would make same sense too.

**Comment by JaroslavTulach** [comment]
There are four compilation failures [IrUtilsTest](https://github.com/enso-org/enso/actions/runs/9600395555/job/26476490971?pr=10274#step:7:2424).

File: docs/syntax/functions.md [hunk 2]
```diff
@@ -100,15 +100,17 @@ binds the function name. This means that:
...
 with a given atom. Methods are dispatched dynamically based on the type of the
-`this` argument, while functions are not.
+`self` argument, while functions are not.
 
 Methods can be defined in Enso in two ways:
 
 1. **In the Body of a Type:** A function defined in the body of a `type`
    definition is automatically converted to a method on all the atoms defined in
-   the body of that type definition.
+   the body of that type definition. It is call an _instance method_. If the
```
Comment (JaroslavTulach) [inline docs/syntax/functions.md]:
```suggestion
   the body of that type definition. If the function has `self` parameter, it is called an _instance method_. If the
```

File: docs/syntax/functions.md [hunk 3]
```diff
@@ -100,15 +100,17 @@ binds the function name. This means that:
...
+   function defined in the body of a type does not have `self` parameter, it is
```
Comment (JaroslavTulach) [inline docs/syntax/functions.md]:
```suggestion
   function does not have `self` parameter, it is called
```

File: docs/syntax/functions.md [hunk 4]
```diff
@@ -122,19 +124,20 @@ type Maybe a
 
 2. **As an Extension Method:** A function defined _explicitly_ on an atom counts
    as an extension method on that atom. It can be defined on a typeset to apply
-   to all the atoms within that typeset.
+   to all the atoms within that typeset. An _extension_ method can be _static_
```
Comment (JaroslavTulach) [inline docs/syntax/functions.md]: `typeset`!? Looks like a relict of some old terminology. Also _"function defined _explicitly_ on an atom"_ doesn't seem to make any sense anymore. In both cases we should probably talk about **type**.

File: docs/syntax/functions.md [hunk 5]
```diff
@@ -122,19 +124,20 @@ type Maybe a
...
 ```
 
-3. **As a Function with an Explicit `this` Argument:** A function defined with
-   the type of the `this` argument specified to be a type.
+3. **As a Function with an Explicit `self` Argument:** A function defined with
+   the type of the `self` argument specified to be a type.
 
 ```ruby
-floor (this : Number) = case this of
+floor (self : Number) = case this of
```
Comment (JaroslavTulach) [inline docs/syntax/functions.md]:
```suggestion
floor (self : Number) = case self of
```

**Review by JaroslavTulach — APPROVED** [review]
Tests look OK. If CI gets green, feel free to proceed with integration.

File: docs/syntax/functions.md (hunk 4 above)
Comment (Akirathan) [reply docs/syntax/functions.md]: Yes, that is a relic. Updated in 46df67bce940f3b56ced66ebd04ffd3b78239c3a

**Review by 4e6 — APPROVED** [review]

**Pull request merged**